### ISSUE TITLE
Finish <*> cleanup

### ISSUE THREAD

**Issue by treeowl** [haskell/containers#129]
I'd really like to get rid of the partial functions with error messages that should never happen. One part of this is to make a new data type for "thin" trees, something like this:

    data Thin a = Thin {-# UNPACK #-} !Int !(Digit12 a) (Thin (Node a)) !(Digit12 a)

This is easy, and rewriting the spine-stretching code to use this type is easy as well. Redoing the `<*>` and `apty` code is a bit trickier, because there are so many different shapes the top of a `FingerTree` can have. The key features of tree tops, from this perspective, are, roughly speaking, these:

1. The numerical value of the left digit
2. The numerical value of the right digit
3. Whether the middle has zero, one, or more than one element

I'm currently only moderately confident that I can find a clean way to do it, but I'm somewhat hopeful. One question is whether I can do so without harming performance of the immediate indexing test—case analysis of the size field may turn out to be inherently faster than pattern matching on the structure of the tree top and views of its middle.

**Comment by treeowl** [comment]
At long last, I am very close to getting this sorted. Most importantly, I've figured out a cleaner way to deal with stitching the pieces together at the bottom that only requires the `fs` sequence to have at least two elements. This makes everything much more manageable. I still have to fix up some size annotations I broke, and go through one more pass of type modifications, but I expect within a day or two I'll have a version with no partial functions whatsoever that is at least as easy to understand.
